lib/Target/TfheRust/TfheRustEmitter.cpp

```diff
@@ -0,0 +1,234 @@
+
+namespace mlir {
+namespace heir {
+namespace tfhe_rust {
+
+// TODO(https://github.com/google/heir/issues/230): Have a separate pass that
+// topo-sorts the gate ops into levels, and use scf.parallel_for to represent
+// them.
+
+void registerToTfheRustTranslation() {
+  TranslateFromMLIRRegistration reg(
+      "emit-tfhe-rust",
+      "translate the tfhe_rs dialect to Rust code for tfhe-rs",
+      [](Operation *op, llvm::raw_ostream &output) {
+        return translateToTfheRust(op, output);
+      },
+      [](DialectRegistry &registry) {
+        registry.insert<func::FuncDialect, tfhe_rust::TfheRustDialect,
+                        arith::ArithDialect>();
+      });
+}
+
+LogicalResult translateToTfheRust(Operation *op, llvm::raw_ostream &os) {
+  SelectVariableNames variableNames(op);
+  TfheRustEmitter emitter(os, &variableNames);
+  LogicalResult result = emitter.translate(*op);
+  return result;
+}
+
+LogicalResult TfheRustEmitter::translate(Operation &op) {
+  LogicalResult status =
+      llvm::TypeSwitch<Operation &, LogicalResult>(op)
+          // Builtin ops
+          .Case<ModuleOp>([&](auto op) { return printOperation(op); })
+          // Func ops
+          .Case<func::FuncOp, func::ReturnOp>(
+              [&](auto op) { return printOperation(op); })
+          // Arith ops
+          .Case<arith::ConstantOp>([&](auto op) { return printOperation(op); })
+          // TfheRust ops
+          .Case<AddOp, ApplyLookupTableOp, ScalarLeftShiftOp>(
+              [&](auto op) { return printOperation(op); })
+          .Default([&](Operation &) {
+            return op.emitOpError("unable to find printer for op");
+          });
+
+  if (failed(status)) {
+    op.emitOpError(llvm::formatv("Failed to translate op {0}", op.getName()));
+    return failure();
+  }
+  return success();
+}
+
+std::string commaSeparatedValues(
+    ValueRange values, std::function<std::string(Value)> valueToString) {
+  return std::accumulate(
+      std::next(values.begin()), values.end(), valueToString(values[0]),
+      [&](std::string &a, Value b) { return a + ", " + valueToString(b); });
+}
+
+FailureOr<std::string> commaSeparatedTypes(
+    TypeRange types, std::function<FailureOr<std::string>(Type)> typeToString) {
+  return std::accumulate(
+      std::next(types.begin()), types.end(), typeToString(types[0]),
+      [&](FailureOr<std::string> &a, Type b) -> FailureOr<std::string> {
+        auto result = typeToString(b);
+        if (failed(result)) {
+          return failure();
+        }
+        return a.value() + ", " + result.value();
+      });
+}
+
+LogicalResult TfheRustEmitter::printOperation(ModuleOp moduleOp) {
+  os << kModulePrelude << "\n";
+  for (Operation &op : moduleOp) {
+    if (failed(translate(op))) {
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+LogicalResult TfheRustEmitter::printOperation(func::FuncOp funcOp) {
+  os << "pub fn " << funcOp.getName() << "(\n";
```

Comment:
kinda silly, but should the func private visibility attribute be used to determine whether the rust fn is `pub`?

Reply:
I suppose we could.... I'm not entirely sure what that would do for us though. I'll think about it and add it to my TODO list.